Handle pid collisions safely in ActiveRecord

cliffcrosland opened the pull request:
### Summary

Due to pid recycling, it is possible for a process to have the same pid as its
grandparent's. A map from pid to connection pools is kept in
`ActiveRecord::ConnectionAdapters::ConnectionHandler`. If a process has the
same pid as its grandparent, then we might fetch a connection pool from this
map that has already been discarded. This can lead to errors when
`#establish_connection` is called in a process whose pid is the same as its
grandparent's.

In addition to discarding the connection pools belonging to an ancestor
process, we should also delete them from the pid map to avoid the pid collision
danger in descendant processes.

Fixes #33022.

cliffcrosland committed:
Handle pid collisions safely in ActiveRecord

Due to pid recycling, it is possible for a process to have the same pid as its
grandparent's. A map from pid to connection pools is kept in
`ActiveRecord::ConnectionAdapters::ConnectionHandler`. If a process has the
same pid as its grandparent, then we might fetch a connection pool from this
map that has already been discarded. This can lead to errors when
`#establish_connection` is called in a process whose pid is the same as its
grandparent's.

In addition to discarding the connection pools belonging to an ancestor
process, we should also delete them from the pid map to avoid the pid collision
danger in descendant processes.

Fixes #33022.

cliffcrosland commented:
Hmm. This interacts badly with the way `#retrieve_connection_pool` looks for connection specifications from ancestor processes.

matthewd commented:
The pid/pool map has always felt a bit heavyweight to me... maybe I'm forgetting a behaviour it provides, but off the top of my head I would guess we only need to store 1) the current process's pid, and 2) the current process's pool.

Then instead of the hash access, we check the pid matches -- if it does, proceed using the existing pool; if it doesn't, store the new pid and replace the pool with a new one, copying the config from the old one at the same time.

@cliffcrosland does that make any sense? I haven't reread the relevant code, so may just be rambling.

cliffcrosland commented:
@matthewd That sounds very reasonable to me. Using the approach you propose would allow us to use `O(1)` memory instead of `O(n)` if we have a long chain of `n` forks.

matthewd commented:
It's worth noting it doesn't help us (doesn't make things worse, but also doesn't make things better) if our grandparent used AR, then our parent did not, and then we do -- and we have received our grandparent's pid. But without proactively detecting the fork, I don't think that's solvable.

cliffcrosland committed:
Use constant memory instead of pid to pools map

cliffcrosland commented:
Great point. Would be a very tricky case to cover. For now, I've added a commit to replace the pid/pool map with two instance variables: one to track the current pid, and one to hold the current pools. If we try to access the pools, but the pid has changed, we reinitialize the pools from the ancestor specifications and discard the ancestor connection pools.

cliffcrosland committed:
Keep configs around, but not connections

cliffcrosland committed:
Use ancestor pool map, smaller change

cliffcrosland commented:
@matthewd Just pushed up another commit. The earlier approach did ensure that we used a constant amount of memory no matter how deep the fork chain was, but it had a few weaknesses:
- The changes were fairly dramatic.
- The fork detection code was not concurrency safe.
- I discovered that eagerly establishing all of the connections from the parent process led to recursive loops.

In this new commit, the change is much smaller. The functionality is very similar to what already exists, with the exception that discarded ancestor connection pools are removed from `@owner_to_pool` and added to a separate `@ancestor_to_pool` map.

cliffcrosland committed:
Fixup space character

SamSaffron commented:
I have been looking at some related issues on Discourse with our custom "pool drainer" https://github.com/SamSaffron/performance/tree/master/pool_drainer 

@tgxworld noticed that after forking we had another ConnectionPool in a discarded state hanging about in the object space.

I did some analysis here per: https://github.com/SamSaffron/performance/blob/master/pool_drainer/draining.rb 

What I am noticing is that the pool is keeping alive the connection reaper thread object (thread is dead) and in turn the connection handler is keeping the pool alive . 

I really like @matthewd idea here https://github.com/rails/rails/pull/33057#issuecomment-394702572 cause then as soon as we open a new connection we will clean up the mess left from the ancestor and don't need to keep a full hash , just a @this_is_the_pid_i_think_i_am var which if we are wrong we clean up. 

Another big long term goal here is to push on Ruby core to add a "after_fork" hook of sorts so we can just do the cleanup then.

cliffcrosland commented:
@SamSaffron That sounds good to me. Maybe I can push up another commit that looks like this:
- Maintain a `@pid` instance variable. 
- In `owner_to_pool`, when we detect that the current pid differs from `@pid`, we discard the ancestor connections and garbage collect them.
- `retrieve_connection_pool` currently looks up spec information from discarded ancestor connection pools in order to establish new connections in the current process. Perhaps when we discard ancestor connections, we can copy the `connection_pool.spec.to_hash` and `connection_pool.schema_cache` data into a separate instance variable. This would allow `retrieve_connection_pool` to use the spec info from ancestor processes without keeping discarded ancestor connection pools around in memory.

SamSaffron commented:
Sounds good, but I am mixed on keeping schema cache, on one hand it is more
efficient, on the other hand we get some extra resilience but re-booting it
on fork, I guess the smaller change is just keeping it

On Sat, 9 Jun 2018 at 3:02 am, Cliff Crosland <notifications@github.com>
wrote:

> @SamSaffron <https://github.com/SamSaffron> That sounds good to me. Maybe
> I can push up another commit that looks like this:
>
>    - Maintain a @pid instance variable.
>    - In owner_to_pool, when we detect that the current pid differs from
>    @pid, we discard the ancestor connections and garbage collect them.
>    - retrieve_connection_pool currently looks up spec information from
>    discarded ancestor connection pools in order to establish new connections
>    in the current process. Perhaps when we discard ancestor connections, we
>    can copy the connection_pool.spec.to_hash and
>    connection_pool.schema_cache data into a separate instance variable.
>    This would allow retrieve_connection_pool to use the spec info from
>    ancestor processes without keeping discarded ancestor connection pools
>    around in memory.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/33057#issuecomment-395823610>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAUXfp6_ef2Koj7UdTEtOHminuAq6owks5t6q44gaJpZM4UZ_IP>
> .
>

eugeneius commented:
There are two distinct issues being discussed here:

1. the new pool discarding behaviour in 5.2 causes a crash when a pid collision occurs
2. connection pools and reapers from ancestor processes are never garbage collected

afaict (2) is not a new problem in 5.2 (@SamSaffron, please correct me if I'm wrong), whereas (1) is a regression. I think we should make the smallest, most backportable change that solves (1), and then fix (2) separately.

Here's my proposed solution:

```patch
diff --git a/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
index c730584902..4081565194 100644
--- a/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
+++ b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
@@ -1034,12 +1034,14 @@ def retrieve_connection_pool(spec_name)
         owner_to_pool.fetch(spec_name) do
           # Check if a connection was previously established in an ancestor process,
           # which may have been forked.
-          if ancestor_pool = pool_from_any_process_for(spec_name)
+          if ancestor_owner_to_pool = owner_to_pool_from_any_process_for(spec_name)
+            ancestor_pool = ancestor_owner_to_pool[spec_name]
             # A connection was established in an ancestor process that must have
             # subsequently forked. We can't reuse the connection, but we can copy
             # the specification and establish a new connection with it.
             establish_connection(ancestor_pool.spec.to_hash).tap do |pool|
               pool.schema_cache = ancestor_pool.schema_cache if ancestor_pool.schema_cache
+              ancestor_owner_to_pool.delete(spec_name)
             end
           else
             owner_to_pool[spec_name] = nil
@@ -1053,9 +1055,8 @@ def owner_to_pool
           @owner_to_pool[Process.pid]
         end

-        def pool_from_any_process_for(spec_name)
-          owner_to_pool = @owner_to_pool.values.reverse.find { |v| v[spec_name] }
-          owner_to_pool && owner_to_pool[spec_name]
+        def owner_to_pool_from_any_process_for(spec_name)
+          @owner_to_pool.values.reverse.find { |v| v[spec_name] }
         end
     end
   end
```

Once we establish a connection pool in the child process, release the ancestor pool that was used to create it. This reuses the existing `owner_to_pool` structure, but never retains more than one instance of a given connection pool.

tgxworld commented:
@eugeneius I tried out your change locally and it does allow the pool to be garbage collected since the map doesn't hold any reference to it. I have a [PR up](https://github.com/rails/rails/pull/33094) for stopping the reaper so (2) should be solved soonish.

cliffcrosland committed:
Use less invasive change, add `discarded?` check

- Use the smaller change proposed by @eugeneius. It is simpler and much less
  invasive.
- Add a check to ensure that we do not call `disconnect!` on a connection pool
  that has already been discarded. This prevents the regression that occurs
  when a process shares the same pid as its grandparent and calls
  `establish_connection`.

cliffcrosland committed:
Replace `discarded?` with ancestor pool removal

cliffcrosland commented:
@eugeneius - I've added your proposed change along with an additional check to avoid calling `disconnect!` on discarded connection pools in `remove_connection`. 

The additional check addresses the failure scenario we're experiencing. Brief summary:

1. The grandparent process initializes connection pools and forks the parent process.
2. The parent process calls `establish_connection` and forks the child process before it has made any calls to `retrieve_connection_pool`. Hence, discarded connection pools from the grandparent still appear in the `@owner_to_pools` map.
3. The child process is assigned the same pid as its grandparent and calls `establish_connection`. This calls `remove_connection`, retrieving the discarded connection pool from the grandparent. When we call `disconnect!` on this discarded connection pool, we experience the errors described in #33022. The additional check avoids making this `disconnect!` call.

However, I think it would probably be better if something in `establish_connection` would remove ancestor connection pools in a way that's similar to what `retrieve_connection_pool` does. Otherwise, if all of the processes in a long fork chain call `establish_connection` without calling `retrieve_connection_pool`, then `@owner_to_map` will continue to bloat over time.

cliffcrosland commented:
@eugeneius - Perhaps this might work better. This way, whether `retrieve_connection_pool` or `establish_connection` is called first, the ancestor pool will be removed.

```diff
diff --git a/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
index c730584..bc44849 100644
--- a/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
+++ b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
@@ -1020,6 +1020,9 @@ def connected?(spec_name)
       # can be used as an argument for #establish_connection, for easily
       # re-establishing the connection.
       def remove_connection(spec_name)
+        @owner_to_pool.each do |pid, pools|
+          pools.delete(spec_name) if pid != Process.pid
+        end
         if pool = owner_to_pool.delete(spec_name)
           pool.automatic_reconnect = false
           pool.disconnect!
@@ -1034,12 +1037,14 @@ def retrieve_connection_pool(spec_name)
         owner_to_pool.fetch(spec_name) do
           # Check if a connection was previously established in an ancestor process,
           # which may have been forked.
-          if ancestor_pool = pool_from_any_process_for(spec_name)
+          if ancestor_owner_to_pool = owner_to_pool_from_any_process_for(spec_name)
+            ancestor_pool = ancestor_owner_to_pool[spec_name]
             # A connection was established in an ancestor process that must have
             # subsequently forked. We can't reuse the connection, but we can copy
             # the specification and establish a new connection with it.
             establish_connection(ancestor_pool.spec.to_hash).tap do |pool|
               pool.schema_cache = ancestor_pool.schema_cache if ancestor_pool.schema_cache
+              ancestor_owner_to_pool.delete(spec_name)
             end
           else
             owner_to_pool[spec_name] = nil
@@ -1053,9 +1058,8 @@ def owner_to_pool
           @owner_to_pool[Process.pid]
         end

-        def pool_from_any_process_for(spec_name)
-          owner_to_pool = @owner_to_pool.values.reverse.find { |v| v[spec_name] }
-          owner_to_pool && owner_to_pool[spec_name]
+        def owner_to_pool_from_any_process_for(spec_name)
+          @owner_to_pool.values.reverse.find { |v| v[spec_name] }
         end
     end
   end
```

cliffcrosland commented:
Hmmm. The diff above will not work. When we reach the `@owner_to_pool.each` block in `remove_connection`, the ancestor connection pools have not yet been discarded because `#owner_to_pool` has not been called yet, which normally triggers the `discard_unowned_pools` method. Hence, when we call `pools.delete(spec_name)`, we will call finalizers on connection pools that haven't been discarded yet, which will kill connections in any ancestor processes that are still running.

Thus, we need to make sure `#owner_to_pools` is called first. This diff should be better:
```diff
diff --git a/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
index c730584..054dd16 100644
--- a/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
+++ b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
@@ -1020,7 +1020,11 @@ def connected?(spec_name)
       # can be used as an argument for #establish_connection, for easily
       # re-establishing the connection.
       def remove_connection(spec_name)
-        if pool = owner_to_pool.delete(spec_name)
+        pool = owner_to_pool.delete(spec_name)
+        if ancestor_owner_to_pool = owner_to_pool_from_any_process_for(spec_name)
+          ancestor_owner_to_pool.delete(spec_name)
+        end
+        if pool
           pool.automatic_reconnect = false
           pool.disconnect!
           pool.spec.config
@@ -1034,12 +1038,14 @@ def retrieve_connection_pool(spec_name)
...
```
I'll push up a commit with this change.

eugeneius commented:
Oh, right. I had incorrectly assumed that after a fork, connection pools would always be established based on the ones from the parent process, but creating entirely new pools needs to work correctly too. It would also be helpful if I actually ran the test case I'm trying to fix. 🙄 

I like the idea of releasing ancestor pools in `remove_connection` - what do you think of this approach, mostly for the benefit of looking more like the corresponding code in `retrieve_connection_pool`:

```patch
diff --git a/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
index c730584902..6dbfe5ae4c 100644
--- a/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
+++ b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
@@ -1024,6 +1024,8 @@ def remove_connection(spec_name)
           pool.automatic_reconnect = false
           pool.disconnect!
           pool.spec.config
+        elsif ancestor_owner_to_pool = owner_to_pool_from_any_process_for(spec_name)
+          ancestor_owner_to_pool.delete(spec_name)
         end
       end
```

eugeneius commented:
Hah - I didn't see your comment before posting mine, but it looks we both arrived at the same idea. 😄

cliffcrosland committed:
Discard ancestor pools when deleting from map

Ancestor pools need to be discarded before their finalizers are called.
Otherwise, if a pool's finalizer is called in a child process, the
pool's connections will be killed in the parent process.

This commit addresses the following scenario.

1. Grandparent process initializes some connection pools. Forks parent
   process.
2. Parent process discards the grandparent's connection pools properly
   because `@owner_to_pool[Process.pid]` creates a new map entry. Forks
   child process.
3. Due to pid recycling, the child process is assigned same pid as
   its grandparent. As a result, `@owner_to_pool[Process.pid]` does not
   create a new map entry, so parent connection pools are not discarded.
4. If we delete the parent connection pools from `@owner_to_pool`, their
   finalizers will run, closing their connections. If the parent process
   is still running, its database connections will be dead.

cliffcrosland commented:
That sounds great. I made the change above, but it looks like there is one more issue to address. Normally, when a process forks, `@owner_to_pool[Process.pid]` will create a new entry in the map, triggering `discard_unowned_pools`. However, if a process's pid collides with its grandparent's, then `@owner_to_pool[Process.pid]` will not create a new entry, and `discard_unowned_pools` will not be called. As a result, the parent connection pools will not be discarded.

In that situation, when we call `ancestor_owner_to_pool.delete(spec_name)`, whether in `remove_connection` or `retrieve_connection_pool`, we might delete a connection pool from our parent that has not yet been discarded. When that happens, the finalizers are called on the connections in that pool, killing the database connections in the parent if it is still running.

I can think of two solutions. Which one do you think is best?

Solution 1. Add `discard!` whenever we delete an ancestor connection pool from a map.

```diff
diff --git a/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
index c730584..a4f1223 100644
--- a/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
+++ b/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb
@@ -1024,6 +1024,8 @@ def remove_connection(spec_name)
           pool.automatic_reconnect = false
           pool.disconnect!
           pool.spec.config
+        elsif ancestor_owner_to_pool = owner_to_pool_from_any_process_for(spec_name)
+          ancestor_owner_to_pool.delete(spec_name).discard!
         end
       end

@@ -1034,12 +1036,15 @@ def retrieve_connection_pool(spec_name)
         owner_to_pool.fetch(spec_name) do
           # Check if a connection was previously established in an ancestor process,
           # which may have been forked.
-          if ancestor_pool = pool_from_any_process_for(spec_name)
+          if ancestor_owner_to_pool = owner_to_pool_from_any_process_for(spec_name)
+            ancestor_pool = ancestor_owner_to_pool[spec_name]
+            ancestor_pool.discard!
             # A connection was established in an ancestor process that must have
             # subsequently forked. We can't reuse the connection, but we can copy
             # the specification and establish a new connection with it.
             establish_connection(ancestor_pool.spec.to_hash).tap do |pool|
               pool.schema_cache = ancestor_pool.schema_cache if ancestor_pool.schema_cache
+              ancestor_owner_to_pool.delete(spec_name)
             end
           else
```

Solution 2. Proactively discard all unowned connections whenever we detect that our pid is different from our parent's pid.
```diff
       private

         def owner_to_pool
+          discard_unowned_connections if Process.pid != @pid
           @owner_to_pool[Process.pid]
         end

+        def discard_unowned_connections
+          ConnectionHandler.discard_unowned_connections(@owner_to_pool)
+          @pid = Process.pid
+        end
```

I think I prefer Solution 1, but I'm open to either.

cliffcrosland committed:
Simplify remove_connection

SamSaffron commented:
@cliffcrosland I wonder aren't both solutions going to be a little bit prone if you do quick double forking? essentially I feel like a lot of a "100% fix here" is going to depend on https://bugs.ruby-lang.org/issues/5446 which both @matthewd  and @tenderlove  are possibly looking to get into AS which paves the road to being able to handle this very cleanly going forward. 

I wonder if we should just pick the simpler solution for now and (2) feels a bit simpler to grok to me. 

Another "trick" I have used elsewhere which can resolve this even in "fast fork" situations is looking at the pool reaper thread. If for any reason it is not running your are safe to clean up the pool and copy the spec. If we "rely" on it running we don't even need to look at PIDs (if we are super careful)

matthewd commented:
> Another "trick" I have used elsewhere which can resolve this even in "fast fork" situations is looking at the pool reaper thread. If for any reason it is not running your are safe to clean up the pool and copy the spec. If we "rely" on it running we don't even need to look at PIDs (if we are super careful)

That sounds interesting. I'm somewhat inclined to spend an extra thread on it, even: just spawn a `Thread.new { sleep 86400 while true }`, then use its liveness for fork detection (without any risk that a reaper exception, or similar, could mislead us).

ktpdmjyet commented:
I am facing the same issue in my project as well. @cliffcrosland how did you proceed finally?
